src/middlewared/middlewared/plugins/fcport.py

```diff
@@ -1,9 +1,382 @@
+                        npiv_setting = fc_hosts[fc_host_alias]['npiv']
+                        if int(chan) > npiv_setting:
+                            verrors.add(
+                                f'{schema_name}.port',
+                                f'Invalid FC port ({port}) supplied, chan {chan} is greater than NPIV setting {npiv_setting}'
+                            )
+                        elif int(chan) == 0:
+                            verrors.add(
+                                f'{schema_name}.port',
+                                f'Invalid FC port ({port}) supplied, chan {chan} must be greater than 0 and less than NPIV setting {npiv_setting}'
+                            )
+                    else:
+                        verrors.add(
+                            f'{schema_name}.port',
+                            f'Invalid FC port ({port}) supplied, digits only after "{VPORT_SEP_CHAR}"'
+                        )
+            else:
+                verrors.add(
+                    f'{schema_name}.port',
+                    f'Invalid FC port ({port}) supplied, should be one of: {",".join(fc_hosts.keys())}'
+                )
+        else:
+            verrors.add(
+                f'{schema_name}.port',
+                f'No FC port supplied, should be based on one of: {",".join(fc_hosts.keys())}'
+            )
+
+        # Test the target supplied
+        target_id = data.get('target_id')
+        if target_id is not None:
+            try:
+                target = await self.middleware.call('iscsi.target.query', [['id', '=', target_id]], {'get': True})
+            except MatchNotFound:
+                verrors.add(
+                    f'{schema_name}.target_id',
+                    f'No target exists with the specified id {target_id}'
+                )
+            else:
+                target_mode = target['mode']
+                if target_mode not in ['FC', 'BOTH']:
+                    target_name = target['name']
+                    verrors.add(
+                        f'{schema_name}.target_id',
+                        f'Specified target "{target_name}" ({target_id}) does not have a "mode" ({target_mode}) that permits FC access'
+                    )
+
+        verrors.check()
+
+    @private
+    async def compress(self, data):
+        for key in ['wwpn', 'wwpn_b']:
+            if key in data:
+                data.pop(key)
+        return data
+
+    @private
+    async def extend(self, data):
+        data['wwpn'] = None
+        data['wwpn_b'] = None
+        # We want to retrieve the WWPN associated with the port name
+        if VPORT_SEP_CHAR in data['port']:
+            # NPIV - Get the base FC Host
+            fc_host_alias, chan = data['port'].rsplit(VPORT_SEP_CHAR, 1)
+        else:
+            fc_host_alias = data['port']
+            chan = None
+        try:
+            fc_host_pair = await self.middleware.call('fc.fc_host.query', [['alias', '=', fc_host_alias]], {'get': True})
```

Comment:
Would it make sense to have a datastore `extend_context` calling `fc.fc_host.query`, then do filter_list here?